presto-spi/src/main/java/com/facebook/presto/spi/block/ImmutableBlock.java

```diff
@@ -0,0 +1,19 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+public interface ImmutableBlock
```

Comment:
@tdcmeehan What's the purpose of the ImmutableBlock? Isn't Block already immutable?

Reply:
The purpose was to provide a parallel type to `BlockBuilder`.

In ASCII art:

```
+-------------+      +--------------+
| BlockBuilder|      |ImmutableBlock|
+------+------+      +--------------+
       ^         +----------^----------+
       |         |                     |
       |         |                     |
       |         |                     |
       |     +---+-+            +------+--------+
       +-----+Block|            | UncheckedBlock|
             +-----+            +---------------+

```

So a `Block` could be a `BlockBuilder`, or it could be an `ImmutableBlock`.  But an `ImmutableBlock` can never be a `BlockBuilder`.  So it's meant to provide some type-safe assurance that using the unchecked methods are safe.

Reply:
@tdcmeehan : Put this ASCII comment into the code would be helpful 😃

Reply:
It was a nice goal to separate out the write path from read-only path.  Unfortunately, the code is messier than the ideal and there's lots of cases where we use a `BlockBuilder` directly as a `Block`.  I'm considering this a separate problem, so to simplify and unify the code, `Block` extends from `UncheckedBlock` now.  This means everything that extends `Block` has the unchecked methods.  As @mbasmanova pointed out, `getUncheckedXXX` methods are already super clear on the fact that these are unchecked.

In other words, no need to include the ASCII art anymore. :)